### PR TITLE
Disk caching attempt...

### PR DESCRIPTION
**Pull request by mhwest13**
I needed to not have facter hitting rightscale every time we decided to load up facter, both inside of ruby and from the facter cli...

This is a crude yet working disk_cache routine / flow that will reduce your rightscale api calls to at the very least some X amount over time as well as across executions or library inclusions... I'm sure this could be implemented more eloquently, I just figured I'd offer this up as a beginning stab so someone might be able to make it better.

Cheers,

p.s. I'm in no way representing myself as a ruby programmer, this took way longer than it probably should have and may have a few more changes than necessary ;)

### CODE REVIEW

**Comment by diranged** [comment]
So sorry we never really commented much on this ... Its been in the back of my mind for a while now. The overall idea is great. We see a few API failures a day running this plugin in production on a few hundred hosts, and they're annoying. It would be great to cache the data so that we have it if necessary.

I havn't merged this particular pull request because it makes a number of assumptions that I would like to  tweak. Additionally you went through and changed all of our uses of symbols into strings, which isn't a huge deal but its a bit of a style nit of mine. Symbols are generally faster, and I think they're easier to read.

That said, I'm going to take this work and in the next week or two I'll probably integrate it in some way into the code so that there is a reliable failsafe. The big change that needs to happen is that there needs to be some wrapping of our `get` methods so that on-API-exceptions, the cache is used. I don't want to use the cache unless a failure has occurred.

**Comment by mhwest13** [comment]
Hey no worries, we've not gone full tilt yet, but we know that once we do, we'll use this and wait for a better approach.

As far as accepting this diff, I was actually hoping someone else would "fix" it because I'm sure I did things that didn't need to happen, or that can happen in faster ways; ruby is not really a language I'm familiar with.

Cheers

**Comment by ryanoleary** [comment]
Just FYI - this code caused massive increased load on the RS API servers that caused alerts and auto-scaling to happen. I'd like to request that it not be merged until that gets figured out. We'll try to take a look here and see if can find the cause.

File: lib/facter/rs-facts.rb [hunk 1]
```diff
@@ -198,20 +237,26 @@ def get_data(data)
...
       # catch new tags that have been created.
       Facter.add(key) {
-        setcode { get_data(:tags)[key] }
+        setcode { get_data('tags')[key] }
+      }
+    end
+  else
+    cache['tags'].each do |k,v|
+      Facter.add(k) {
+        setcode { get_data('tags')[k] }
```
Comment (ryanoleary) [inline lib/facter/rs-facts.rb]: This is problematic -- the recursive call to get_data will happen forever.

**Comment by cdenneen** [comment]
Anything further on this "fix"?

**Pull request closed without merging**